usr/src/uts/common/fs/zfs/zfs_vfsops.c

```diff
@@ -1023,6 +1023,14 @@ zfsvfs_setup(zfsvfs_t *zfsvfs, boolean_t mounting)
 {
 	int error;
 
+	/*
+	 * Check for a bad on-disk format version now since we
+	 * lied about owning the dataset readonly before.
+	 */
+	if (!(zfsvfs->z_vfs->vfs_flag & VFS_RDONLY) &&
```

Comment:
You can't (reliably) check for VFS_RDONLY until after zfs_register_callbacks has been called.